qi-lib/flow/impl.rkt

```diff
@@ -186,6 +189,9 @@
         (append (values->list (apply op vs))
                 (apply zip-with op (map rest seqs))))))
 
+(define 1->1 (thunk  (values)))
+(define *->1 (thunk* (values)))
```

Comment:
These appear in error messages shown to the user:
```
> (~> (1 2 3) (relay))
; 1->1: arity mismatch;
;  the expected number of arguments does not match the given number
;   expected: 0
;   given: 3
```

Any reason not to use `(procedure-rename ...)` within `relay`, `relay*` and other forms using `1->1` and `*->1` so that the name of the form used (e.g. `relay` in the above example) is reported to the user?

Reply:
> Could you explain the notation here? It doesn't seem to indicate the number of values since the first is 0->0, and the second is N->0.

`Values` can be regard as a kind of Cartesian Product, so that `(values "a" "b" "c")` can be marked as `"a" × "b" × "c"`. In category theory `1` is the identity element of `×`, which is `(values)` in racket.

`*->1` : accepts any arguments and return `(values)`.
`1->1` : accepts no value and return `(values)`.

On the one hand, I haven't thought of a better notation for `(values)`. On the other hand, Qi seems to have a deep connection with category theory, so I think it makes sense to use the notation in category theory directly (and we can use `+` and `0` to represent `covalues` and the identity element of it).

> Any reason not to use (procedure-rename ...) within relay, relay* and other forms using 1->1 and *->1 so that the name of the form used (e.g. relay in the above example) is reported to the user?

For example, `*->1` is not only the identity element of `-<`, but also `⏚` in qi:

```racket
> (eq? (☯ ⏚) (☯ (-<)))
#t
```

And if we rename these procedures, the equality will be lost.

Reply:
Thanks for clarifying! The correspondence to category theory and the duality between sum and product makes sense, but do you anticipate any particular advantage gained by having the `eq?` equivalence? In general since equality of functions is undecidable, I would be skeptical of code that employs logic based on checks for equivalence between functions based on their identities. Unless there are some specific benefits you have in mind, I would favor keeping the names recognizable in error messages to supporting an `eq?` equivalence. We can still have the actual functions named as `1->1` and `*->1` as that would preserve the identity from the perspective of the codebase instead of having duplicate implementations in the different forms.

Reply:
> but do you anticipate any particular advantage gained by having the eq? equivalence? In general since equality of functions is undecidable, I would be skeptical of code that employs logic based on checks for equivalence between functions based on their identities.

Yes, in general functions are undecidable. But `*->1` is special, it is the identity element of `-<` (a monoid), so it should have this property:

```racket
Welcome to Racket v8.6 [cs].
> (require qi)
> (eq? (~> () (-< 123 (-<))) (~> () (-< 123)))
#t
> (eq? (~> () (-< (-<) 123)) (~> () (-< 123)))
#t
```

I prefer to preserve the properties of mathematical structures as much as possible.

Reply:
Yes, I agree that we should aim to preserve mathematical properties. Yet, `eq?` is not a mathematical relation but an implementation-dependent relation in the Scheme world. It doesn't assess equality based on properties of the objects being compared but based on arbitrary details of the implementation (e.g. memory location where the values may happen to be stored). As a result, I would say that we should avoid considering `eq?` behavior in our design process except in cases where it provides a compelling performance benefit in practice -- otherwise, better to think in terms of `equal?` (but even that is problematic -- [the paper on egal?](https://dl.acm.org/doi/pdf/10.1145/165593.165596) covers some of the issues but not all).

Aside from the specific choice of equality relation, in the present case, I feel the mathematical properties we'd like to preserve are:

```
((☯ (-< (-<) f)) arg ...) = (f arg ...) = ((☯ (-< f (-<))) arg ...)
```

That is, an operational equivalence in terms of the result of applying these functions to arguments. Since, for instance, we could have a totally different definition of `*->1` which would also fulfill the monoid laws, but would not be `eq?` to the `*->1` defined in the codebase. And in this case, there is nothing specific that we need to do in order to ensure the above relation holds, as `(relay)` and others would satisfy this relation even without being `eq?`.

Btw, I also meant that code in general should not do checks like `(if (eq? f1 f2) ...)` or even `(if (equal? f1 f2) ...)` or `(if (member f1 (list f2 f3 f3))` where the fs are functions, since in the general case this is undecidable, and in special cases, it is implementation-specific and akin to a "hack" for performance. In this line of thinking, we should not encourage users writing code like `(if (member (☯ (-<)) (list f ...))`. Instead, we could simply apply the function to a relevant argument of interest, `(if (= v (☯ (-<)) v) ...)` without making a general statement that the function "is" the monoid identity.

Reply:
> As a result, I would say that we should avoid considering eq? behavior in our design process except in cases where it provides a compelling performance benefit in practice.

I think `eq?` does provide performance benefit. Because in this way programmers can insert the identity functions anywhere without worrying about the performance overhead.

For example, I thought about simulating the [limit](https://en.wikipedia.org/wiki/Limit_(category_theory)) in category theory by inserting procedures between the arguments of `compose`.
```racket
(define do (make-parameter values))
(~> f g h ...) ; = (~> f (do) g (do) h (do) ...)
```

But it affects the performance of `qi`'s original code -- because it inserts `values`s between all the function arguments of `compose`. This is what motivated me to submit this [PR](https://github.com/racket/racket/pull/4264).


And on the other hand, I'm not sure if it's a good idea to rename the returned procedures in any case. If we rename `(-<)`, how should we deal with `(-< add1)`? If we decide to rename `add1`, there seems to be 2 ways:

```racket
> ((procedure-rename add1 'compiled-tee-flow) 1 2 3)
compiled-tee-flow: arity mismatch;
 the expected number of arguments does not match the given number
  expected: 1
  given: 3
 [,bt for context]
> ((let ([compiled-tee-flow (lambda args (apply add1 args))]) compiled-tee-flow) 1 2 3)
add1: arity mismatch;
 the expected number of arguments does not match the given number
  expected: 1
  given: 3
 [,bt for context]
```

The 1st way might be consistent with the way you expect to rename `*->1`, and the 2nd way is consistent with the case that `-<` accepts more arguments:

```racket
> (~> (1 2 3) (-< add1 sub1))
add1: arity mismatch;
 the expected number of arguments does not match the given number
  expected: 1
  given: 3
 [,bt for context]
```

I'm not sure if renaming named functions (like `add1`, `*->1`) can make it easier for programmers to debug.